File: frontend/src/types/estimate.ts
export type EstimateStatusValue = 'DRAFT' | 'FINAL' | 'ACCEPTED' | 'ARCHIVED';
export type LineOriginValue = 'PLANNED_WORK' | 'MANUAL';
export type QuantitySourceValue = 'SURFACE_NET_AREA' | 'REVEAL_LENGTH' | 'REVEAL_AREA' | 'MANUAL';

export interface EstimateSummaryRead {
  id: string;
  project_id: string;
  version: number;
  status: EstimateStatusValue;
  name: string | null;
  total: string | null;
  currency: string;
  created_at: string;
  updated_at: string;
}

export interface EstimateListResponse {
  items: EstimateSummaryRead[];
  total: number;
}

export interface EstimateLineRead {
  id: string;
  estimate_id: string;
  origin: LineOriginValue;
  position: number;
  description: string;
  item_code: string | null;
  unit: string;
  scope: string;
  currency: string;
  source_quantity: string | null;
  quantity: string;
  quantity_source: QuantitySourceValue;
  quantity_overridden: boolean;
  unit_price: string | null;
  price_override: boolean;
  amount: string | null;
  price_item_id: string | null;
  plan_id: string | null;
  planned_work_id: string | null;
  surface_id: string | null;
  room_id: string | null;
  opening_id: string | null;
  // Presentation metadata resolved live at read time — optional AND nullable:
  // some response-construction paths may omit these keys entirely rather
  // than send an explicit null, so callers must not assume `!== null` alone
  // rules out `undefined`.
  room_name?: string | null;
  surface_name?: string | null;
  surface_type_value?: string | null;
  opening_name?: string | null;
  opening_type_value?: string | null;
}

export interface EstimateRead extends EstimateSummaryRead {
  lines: EstimateLineRead[];
}

// PATCH body for a DRAFT line (Stage 10E contract). Every field is optional —
// an omitted key means "leave unchanged". `unit_price: null` is a distinct,
// meaningful operation (marks the line Do ustalenia / price_override=true);
// `quantity: null` must never be sent (rejected by the backend — NOT NULL).
// `reset_quantity_override`/`reset_price_override` must not be combined with
// a simultaneous `quantity`/`unit_price` assignment on the same field.
export interface EstimateLineUpdatePayload {
  quantity?: string;
  unit_price?: string | null;
  reset_price_override?: boolean;
  reset_quantity_override?: boolean;
}

// Stage 10G.3B — regeneration preview/confirm (Stage 10E contract). Both
// regenerate-preview and regenerate return this same shape; preview never
// mutates the estimate, regenerate applies it in-place.
export type LineChangeTypeValue = 'ADDED' | 'REMOVED' | 'UPDATED';

export interface LineChangeEntry {
  change_type: LineChangeTypeValue;
  estimate_line_id: string | null;
  planned_work_id: string;
  surface_id: string | null;
  opening_id: string | null;
  item_code: string | null;
  description: string;
  unit: string;
  old_source_quantity: string | null;
  new_source_quantity: string | null;
  old_unit_price: string | null;
  new_unit_price: string | null;
  quantity_overridden: boolean;
  price_override: boolean;
  // Stage 10G.3B provenance follow-up — resolved live from current DB
  // records, analogous to EstimateLineRead's presentation fields. Optional
  // AND nullable: some response-construction paths may omit these keys
  // entirely rather than send an explicit null (10G.2 crash-regression
  // protection), and a REMOVED entry's referenced Surface/Opening may no
  // longer exist, in which case the field is null rather than fabricated.
  room_name?: string | null;
  surface_name?: string | null;
  surface_type_value?: string | null;
  opening_name?: string | null;
  opening_type_value?: string | null;
}

export interface RegenerationPreviewResponse {
  added: number;
  removed: number;
  updated: number;
  preserved_manual: number;
  changes: LineChangeEntry[];
}

// Stage 10G.3C — POST body for a MANUAL EstimateLine (Stage 10E contract).
// scope/unit are plain strings here (matching EstimateLineRead), but the UI
// only ever offers PriceScopeValue/PriceUnitValue options; LABOR_AND_MATERIAL
// is never offered — the backend rejects it for MANUAL lines (no PriceBook
// source to split into labor/material components). currency is optional and,
// when sent, must match the estimate's own currency (Stage 10E contract) —
// callers always send the estimate's authoritative currency, never a
// user-editable value.
export interface ManualLineCreatePayload {
  description: string;
  scope: string;
  unit: string;
  quantity: string;
  unit_price?: string | null;
  currency?: string;
}

File: backend/app/schemas/estimate.py
"""Estimate API schemas (Stage 10E / 10E-corrections).

Money stays Decimal end-to-end — never float.
NULL unit_price ≠ 0.00: nullable Decimal serialises as JSON null.
Enums serialise as their string value (str enum mixin on model enums).
"""
from datetime import datetime
from decimal import Decimal
import uuid

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.estimate import EstimateStatus, LineOrigin, QuantitySource
from app.models.price_item import PriceScope, PriceUnit


# ---------------------------------------------------------------------------
# Estimate list / summary
# ---------------------------------------------------------------------------

class EstimateSummaryRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    version: int
    status: EstimateStatus
    name: str | None = None
    total: Decimal | None = None
    currency: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EstimateListResponse(BaseModel):
    items: list[EstimateSummaryRead]
    total: int


# ---------------------------------------------------------------------------
# Estimate line — full detail (B / G)
# ---------------------------------------------------------------------------

class EstimateLineRead(BaseModel):
    id: uuid.UUID
    estimate_id: uuid.UUID
    origin: LineOrigin
    position: int
    description: str
    item_code: str | None = None
    unit: PriceUnit
    scope: PriceScope
    currency: str
    source_quantity: Decimal | None = None
    quantity: Decimal
    quantity_source: QuantitySource
    quantity_overridden: bool
    unit_price: Decimal | None = None
    price_override: bool
    amount: Decimal | None = None
    # Provenance IDs (snapshot)
    price_item_id: uuid.UUID | None = None
    plan_id: uuid.UUID | None = None
    planned_work_id: uuid.UUID | None = None
    surface_id: uuid.UUID | None = None
    room_id: uuid.UUID | None = None
    opening_id: uuid.UUID | None = None
    # Presentation metadata resolved live at read time (not stored in EstimateLine)
    room_name: str | None = None
    surface_name: str | None = None
    surface_type_value: str | None = None
    opening_name: str | None = None
    opening_type_value: str | None = None

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Estimate detail (single estimate + lines)
# ---------------------------------------------------------------------------

class EstimateRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    version: int
    status: EstimateStatus
    name: str | None = None
    total: Decimal | None = None
    currency: str
    lines: list[EstimateLineRead] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Regeneration preview / confirm — structured change entries (C6)
# ---------------------------------------------------------------------------

class LineChangeEntryRead(BaseModel):
    """One proposed or applied change entry in a regeneration diff."""
    change_type: str  # "ADDED" | "REMOVED" | "UPDATED"
    estimate_line_id: uuid.UUID | None = None
    planned_work_id: uuid.UUID
    surface_id: uuid.UUID | None = None
    opening_id: uuid.UUID | None = None
    item_code: str | None = None
    description: str
    unit: PriceUnit
    old_source_quantity: Decimal | None = None
    new_source_quantity: Decimal | None = None
    old_unit_price: Decimal | None = None
    new_unit_price: Decimal | None = None
    quantity_overridden: bool
    price_override: bool


class RegenerationPreviewResponse(BaseModel):
    added: int
    removed: int
    updated: int
    preserved_manual: int
    changes: list[LineChangeEntryRead] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Manual line creation (F)
# ---------------------------------------------------------------------------

class ManualLineCreate(BaseModel):
    description: str = Field(min_length=1, max_length=255)
    scope: PriceScope
    unit: PriceUnit
    quantity: Decimal
    unit_price: Decimal | None = None
    currency: str = Field(default="PLN", max_length=3)

    model_config = ConfigDict(extra="forbid")


# ---------------------------------------------------------------------------
# Line PATCH (G) — update owner-controlled fields
# ---------------------------------------------------------------------------

class EstimateLineUpdate(BaseModel):
    """PATCH body for a draft line.

    Only fields present in model_fields_set are applied.

    quantity=null is rejected (NOT NULL in DB).
    unit_price=null marks the line as Do ustalenia (price_override=True, unit_price=NULL).
    unit_price=<value> sets a price override.
    reset_price_override=true restores unit_price from the current PriceBook and
        clears price_override. Rejected for MANUAL lines. Cannot be combined with
        a simultaneous unit_price assignment.

    reset_quantity_override=true restores quantity to source_quantity (or 0.000 if
        source_quantity is NULL) and clears quantity_overridden. Cannot be combined
        with a simultaneous quantity assignment.

    description is only writable on MANUAL lines.
    """
    quantity: Decimal | None = None
    unit_price: Decimal | None = None
    description: str | None = None
    reset_price_override: bool = False
    reset_quantity_override: bool = False

    model_config = ConfigDict(extra="forbid")

    @model_validator(mode="after")
    def check_no_override_conflicts(self) -> "EstimateLineUpdate":
        if "unit_price" in self.model_fields_set and self.reset_price_override:
            raise ValueError(
                "Cannot set unit_price and reset_price_override simultaneously"
            )
        if "quantity" in self.model_fields_set and self.reset_quantity_override:
            raise ValueError(
                "Cannot set quantity and reset_quantity_override simultaneously"
            )
        return self
